syntax = "proto3";

service RaftService {
    rpc RequestVote(RequestVoteMessage) returns (RequestVoteResponse);

    rpc AppendEntries(AppendEntriesMessage) returns (AppendEntriesResponse);

    rpc GetLeader(GetLeaderMessage) returns (GetLeaderResponse);

    rpc Suspend(SuspendMessage) returns (SuspendResponse);
}

message RequestVoteMessage{
    int32 term = 1;
    int32 id = 2;
}

message RequestVoteResponse{
    int32 term = 1;
    bool vote = 2;
}

message AppendEntriesMessage {
    int32 term = 1;
    int32 id = 2;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool ack = 2;
}

message GetLeaderMessage{
    string message  = 1;
}

message GetLeaderResponse{ 
    uint32 id = 1;
    string address = 2;
}

message SuspendMessage{
    uint32 timeout = 1;
}

message SuspendResponse{}